### PR TITLE
[SYCL][Bindless][Doc] Add support for cubemaps

### PR DESCRIPTION
**Pull request by Seanst98**
Add cubemap support:
 - Allocation and freeing of cubemapped images
 - Unsampled fetching and writing, and sampled reading
 - Device queries for cubemap support
 - Testing for both unsampled and sampled cubemap examples
 - Update the spec with cubemap support

Remove `const` and `&` qualifiers from spec and implementation for handle parameters in `write_xxx` functions.

Corresponding UR PR: https://github.com/oneapi-src/unified-runtime/pull/1433

### CODE REVIEW

**Review by isaacault — APPROVED** [review]

**Comment by Seanst98** [comment]
Friendly ping @intel/dpcpp-tools-reviewers, @intel/unified-runtime-reviewers, @intel/llvm-reviewers-runtime, @intel/llvm-reviewers-cuda, @maarquitos14, and @jchlanda for reviews/approvals

File: sycl/test-e2e/bindless_images/cubemap/cubemap_unsampled.cpp
Comment (JackAKirk) [inline sycl/test-e2e/bindless_images/cubemap/cubemap_unsampled.cpp:78]: unused variable

**Review by DBDuncan — APPROVED** [review]

**Review by AlexeySachkov — APPROVED** [review]
`DeviceConfigFile.td` changes LGTM

File: sycl/test-e2e/bindless_images/cubemap/cubemap_unsampled.cpp
Comment (Seanst98) [reply sycl/test-e2e/bindless_images/cubemap/cubemap_unsampled.cpp]: You're right. Removed `ctxt`.

**Review by jchlanda — APPROVED** [review]
libclc 👍

**Comment by Seanst98** [comment]
Friendly ping @JackAKirk, @intel/unified-runtime-reviewers, @intel/llvm-reviewers-runtime, and @maarquitos14 for reviews/approvals

**Comment by JackAKirk** [comment]
> Friendly ping @JackAKirk, @intel/unified-runtime-reviewers, @intel/llvm-reviewers-runtime, and @maarquitos14 for reviews/approvals

Jakub already approved it for cuda reviewers. You don't need my approval also.

File: sycl/include/sycl/detail/pi.h
Comment (maarquitos14) [inline sycl/include/sycl/detail/pi.h:159]:
```suggestion
// 15.48 Added cubemap support:
```

File: sycl/include/sycl/ext/oneapi/bindless_images.hpp
Comment (maarquitos14) [inline sycl/include/sycl/ext/oneapi/bindless_images.hpp:686]: I guess this was forced by clang-format and there was no real change here, right? If so, please, return back to original state so that blame can point to the last meaningful change.

File: sycl/include/sycl/ext/oneapi/bindless_images.hpp
Comment (maarquitos14) [inline sycl/include/sycl/ext/oneapi/bindless_images.hpp:1194]: I wonder if this should throw rather than asserting. If someone goes through this code path, they would not receive any error, nor even a warning, but the code would not do what they expect. Any thoughts?

File: sycl/include/sycl/ext/oneapi/bindless_images.hpp
Comment (maarquitos14) [inline sycl/include/sycl/ext/oneapi/bindless_images.hpp:1281]: `imageHandle` is not const in the specification. Considering that `imageHandle` is indeed const in both `fetch_cubemap` and `sample_cubemap`, maybe the error is in the spec rather than here, right?

File: sycl/plugins/unified_runtime/CMakeLists.txt
Comment (maarquitos14) [inline sycl/plugins/unified_runtime/CMakeLists.txt:84]: Just a reminder to bring back the correct repo and tag.

File: sycl/source/detail/device_impl.cpp
Comment (maarquitos14) [inline sycl/source/detail/device_impl.cpp:552]: Is this `nocheck` on purpose? Why does it need to be `nocheck`?

File: sycl/source/detail/device_impl.cpp
Comment (maarquitos14) [inline sycl/source/detail/device_impl.cpp:560]: Is this `nocheck` on purpose? Why does it need to be `nocheck`?

File: sycl/test-e2e/bindless_images/cubemap/cubemap_sampled.cpp
Comment (maarquitos14) [inline sycl/test-e2e/bindless_images/cubemap/cubemap_sampled.cpp:4]: Do multiple `REQUIRES` line work as `&&`? Have you made sure the test is actually running?

File: sycl/test-e2e/bindless_images/cubemap/cubemap_sampled.cpp
Comment (maarquitos14) [inline sycl/test-e2e/bindless_images/cubemap/cubemap_sampled.cpp:12]:
```suggestion
// Uncomment to print additional test information.
```

File: sycl/test-e2e/bindless_images/cubemap/cubemap_sampled.cpp
Comment (maarquitos14) [inline sycl/test-e2e/bindless_images/cubemap/cubemap_sampled.cpp:25]:
```suggestion
  // declare image data.
```

File: sycl/test-e2e/bindless_images/cubemap/cubemap_sampled.cpp
Comment (maarquitos14) [inline sycl/test-e2e/bindless_images/cubemap/cubemap_sampled.cpp:26]:
```suggestion
  // width and height must be equal.
  
```

File: sycl/test-e2e/bindless_images/cubemap/cubemap_sampled.cpp
Comment (maarquitos14) [inline sycl/test-e2e/bindless_images/cubemap/cubemap_sampled.cpp:48]:
```suggestion
  // Extension: image descriptor - Cubemap.
```

File: sycl/test-e2e/bindless_images/cubemap/cubemap_sampled.cpp
Comment (maarquitos14) [inline sycl/test-e2e/bindless_images/cubemap/cubemap_sampled.cpp:59]:
```suggestion
    // Extension: allocate memory on device and create the handle.
```

File: sycl/test-e2e/bindless_images/cubemap/cubemap_sampled.cpp
Comment (maarquitos14) [inline sycl/test-e2e/bindless_images/cubemap/cubemap_sampled.cpp:62]:
```suggestion
    // Extension: create the image and return the handle.
```

File: sycl/test-e2e/bindless_images/cubemap/cubemap_sampled.cpp
Comment (maarquitos14) [inline sycl/test-e2e/bindless_images/cubemap/cubemap_sampled.cpp:66]:
```suggestion
    // Extension: copy over data to device (handler variant).
```

File: sycl/test-e2e/bindless_images/cubemap/cubemap_sampled.cpp
Comment (maarquitos14) [inline sycl/test-e2e/bindless_images/cubemap/cubemap_sampled.cpp:78]:
```suggestion
      // Emanating vector scans one face.
```

File: sycl/test-e2e/bindless_images/cubemap/cubemap_sampled.cpp
Comment (maarquitos14) [inline sycl/test-e2e/bindless_images/cubemap/cubemap_sampled.cpp:100]:
```suggestion
            // Extension: sample cubemap data from handle.
```

File: sycl/test-e2e/bindless_images/cubemap/cubemap_sampled.cpp
Comment (maarquitos14) [inline sycl/test-e2e/bindless_images/cubemap/cubemap_sampled.cpp:109]:
```suggestion
    // Extension: cleanup.
```

File: sycl/test-e2e/bindless_images/cubemap/cubemap_sampled.cpp
Comment (maarquitos14) [inline sycl/test-e2e/bindless_images/cubemap/cubemap_sampled.cpp:119]:
```suggestion
  // Collect and validate output.
```

File: sycl/test-e2e/bindless_images/cubemap/cubemap_unsampled.cpp
Comment (maarquitos14) [inline sycl/test-e2e/bindless_images/cubemap/cubemap_unsampled.cpp:3]: Same than previous file.

File: sycl/test-e2e/bindless_images/cubemap/cubemap_unsampled.cpp
Comment (maarquitos14) [inline sycl/test-e2e/bindless_images/cubemap/cubemap_unsampled.cpp:16]:
```suggestion
// Uncomment to print additional test information.
```

File: sycl/test-e2e/bindless_images/cubemap/cubemap_unsampled.cpp
Comment (maarquitos14) [inline sycl/test-e2e/bindless_images/cubemap/cubemap_unsampled.cpp:21]:
```suggestion
// Helpers and utilities.
```

File: sycl/test-e2e/bindless_images/cubemap/cubemap_unsampled.cpp
Comment (maarquitos14) [inline sycl/test-e2e/bindless_images/cubemap/cubemap_unsampled.cpp:79]:
```suggestion
  // Skip half tests if not supported.
```

File: sycl/test-e2e/bindless_images/cubemap/cubemap_unsampled.cpp
Comment (maarquitos14) [inline sycl/test-e2e/bindless_images/cubemap/cubemap_unsampled.cpp:105]:
```suggestion
    // Extension: allocate memory on device and create the handle.
```

File: sycl/test-e2e/bindless_images/cubemap/cubemap_unsampled.cpp
Comment (maarquitos14) [inline sycl/test-e2e/bindless_images/cubemap/cubemap_unsampled.cpp:114]:
```suggestion
    // Extension: copy over data to device.
```

File: sycl/test-e2e/bindless_images/cubemap/cubemap_unsampled.cpp
Comment (maarquitos14) [inline sycl/test-e2e/bindless_images/cubemap/cubemap_unsampled.cpp:127]:
```suggestion
    // Cleanup.
```

File: sycl/test-e2e/bindless_images/cubemap/cubemap_unsampled.cpp
Comment (maarquitos14) [inline sycl/test-e2e/bindless_images/cubemap/cubemap_unsampled.cpp:139]:
```suggestion
  // Collect and validate output.
```

File: sycl/test-e2e/bindless_images/image_get_info.cpp
Comment (maarquitos14) [inline sycl/test-e2e/bindless_images/image_get_info.cpp:100]:
```suggestion
    // Extension: query for bindless image cubemaps support -- aspects.
```

File: sycl/include/sycl/ext/oneapi/bindless_images.hpp
Comment (Seanst98) [reply sycl/include/sycl/ext/oneapi/bindless_images.hpp:1192]: This is kernel code. I didn't think you could throw in kernel code. Am I wrong?

File: sycl/plugins/unified_runtime/CMakeLists.txt
Comment (Seanst98) [reply sycl/plugins/unified_runtime/CMakeLists.txt:84]: Unified Runtime won't approve this PR until the tag has been updated to point towards the latest of UR. Only once it does will this PR gain approval from UR and merged in.

File: sycl/source/detail/device_impl.cpp
Comment (Seanst98) [reply sycl/source/detail/device_impl.cpp:552]: This is what is used elsewhere in this file to query for device aspects so I'm just following the norm. 

However, if I was to guess, it would be that failing a query for a device aspect is not a critical failure and so reporting a `false` is appropriate.

File: sycl/source/detail/device_impl.cpp
Comment (Seanst98) [reply sycl/source/detail/device_impl.cpp:560]: See above thread [here](https://github.com/intel/llvm/pull/12996#discussion_r1535455156).

File: sycl/test-e2e/bindless_images/cubemap/cubemap_sampled.cpp
Comment (Seanst98) [reply sycl/test-e2e/bindless_images/cubemap/cubemap_sampled.cpp:2]: I can't combine the `REQUIRES` lines with `&&` but a simple `,` seems to work. I've switched to that where it fits in a line. Yes the test is running :smile:

File: sycl/test-e2e/bindless_images/cubemap/cubemap_sampled.cpp
Comment (Seanst98) [reply sycl/test-e2e/bindless_images/cubemap/cubemap_sampled.cpp:98]: Done.

File: sycl/test-e2e/bindless_images/image_get_info.cpp
Comment (Seanst98) [reply sycl/test-e2e/bindless_images/image_get_info.cpp]: Done.

File: sycl/include/sycl/detail/pi.h
Comment (Seanst98) [reply sycl/include/sycl/detail/pi.h]: Done.

File: sycl/include/sycl/ext/oneapi/bindless_images.hpp
Comment (Seanst98) [reply sycl/include/sycl/ext/oneapi/bindless_images.hpp]: Yes, good point. Reverted.

File: sycl/include/sycl/ext/oneapi/bindless_images.hpp
Comment (Seanst98) [reply sycl/include/sycl/ext/oneapi/bindless_images.hpp]: Good catch. There's quite a bit of inconsistency with how we're applying `const` to these handles across the spec and implementation. When it comes to writing, the handles should not be `const`. I've applied a fix to the cubemaps funcs and elsewhere in the spec and impl where necessary.

File: sycl/test-e2e/bindless_images/cubemap/cubemap_sampled.cpp
Comment (Seanst98) [reply sycl/test-e2e/bindless_images/cubemap/cubemap_sampled.cpp]: Done.

File: sycl/test-e2e/bindless_images/cubemap/cubemap_sampled.cpp
Comment (Seanst98) [reply sycl/test-e2e/bindless_images/cubemap/cubemap_sampled.cpp]: Done.

File: sycl/test-e2e/bindless_images/cubemap/cubemap_sampled.cpp
Comment (Seanst98) [reply sycl/test-e2e/bindless_images/cubemap/cubemap_sampled.cpp]: Done.

File: sycl/test-e2e/bindless_images/cubemap/cubemap_sampled.cpp
Comment (Seanst98) [reply sycl/test-e2e/bindless_images/cubemap/cubemap_sampled.cpp]: Done.

File: sycl/test-e2e/bindless_images/cubemap/cubemap_sampled.cpp
Comment (Seanst98) [reply sycl/test-e2e/bindless_images/cubemap/cubemap_sampled.cpp]: Done.

File: sycl/test-e2e/bindless_images/cubemap/cubemap_sampled.cpp
Comment (Seanst98) [reply sycl/test-e2e/bindless_images/cubemap/cubemap_sampled.cpp]: Done.

File: sycl/test-e2e/bindless_images/cubemap/cubemap_sampled.cpp
Comment (Seanst98) [reply sycl/test-e2e/bindless_images/cubemap/cubemap_sampled.cpp]: Done.

File: sycl/test-e2e/bindless_images/cubemap/cubemap_sampled.cpp
Comment (Seanst98) [reply sycl/test-e2e/bindless_images/cubemap/cubemap_sampled.cpp]: Done.

File: sycl/test-e2e/bindless_images/cubemap/cubemap_sampled.cpp
Comment (Seanst98) [reply sycl/test-e2e/bindless_images/cubemap/cubemap_sampled.cpp]: Done.

File: sycl/test-e2e/bindless_images/cubemap/cubemap_sampled.cpp
Comment (Seanst98) [reply sycl/test-e2e/bindless_images/cubemap/cubemap_sampled.cpp]: Done.

File: sycl/test-e2e/bindless_images/cubemap/cubemap_unsampled.cpp
Comment (Seanst98) [reply sycl/test-e2e/bindless_images/cubemap/cubemap_unsampled.cpp]: Done.

File: sycl/test-e2e/bindless_images/cubemap/cubemap_unsampled.cpp
Comment (Seanst98) [reply sycl/test-e2e/bindless_images/cubemap/cubemap_unsampled.cpp]: Done.

File: sycl/test-e2e/bindless_images/cubemap/cubemap_unsampled.cpp
Comment (Seanst98) [reply sycl/test-e2e/bindless_images/cubemap/cubemap_unsampled.cpp]: Done.

File: sycl/test-e2e/bindless_images/cubemap/cubemap_unsampled.cpp
Comment (Seanst98) [reply sycl/test-e2e/bindless_images/cubemap/cubemap_unsampled.cpp]: Done.

File: sycl/test-e2e/bindless_images/cubemap/cubemap_unsampled.cpp
Comment (Seanst98) [reply sycl/test-e2e/bindless_images/cubemap/cubemap_unsampled.cpp]: Done.

File: sycl/test-e2e/bindless_images/cubemap/cubemap_unsampled.cpp
Comment (Seanst98) [reply sycl/test-e2e/bindless_images/cubemap/cubemap_unsampled.cpp]: Done.

File: sycl/test-e2e/bindless_images/cubemap/cubemap_unsampled.cpp
Comment (Seanst98) [reply sycl/test-e2e/bindless_images/cubemap/cubemap_unsampled.cpp]: Done.

File: sycl/test-e2e/bindless_images/cubemap/cubemap_unsampled.cpp
Comment (Seanst98) [reply sycl/test-e2e/bindless_images/cubemap/cubemap_unsampled.cpp]: Done.

File: sycl/include/sycl/ext/oneapi/bindless_images.hpp
Comment (cppchedy) [inline sycl/include/sycl/ext/oneapi/bindless_images.hpp:1207]: this is an ABI break.

File: sycl/include/sycl/ext/oneapi/bindless_images.hpp
Comment (DBDuncan) [reply sycl/include/sycl/ext/oneapi/bindless_images.hpp:1192]: A throw should work as it would only run on the host. So no worries about running it on a device that does not support exceptions. Though the sycl 2020 spec states that exceptions are not supported. So perhaps it is better to put something like `assert(false && "Bindless images not yet implemented on host");` instead.

Unless perhaps throws are used elsewhere in situations similar to this. I don't think this should be an outlier using a throw.

Comment (maarquitos14) [reply sycl/include/sycl/ext/oneapi/bindless_images.hpp:1192]: Sorry, I wasn't clear in my first comment: the problem is that `assert` only runs in debug mode, so if a user is using production mode (the most frequent), the assertion wouldn't work, and the program would continue as if nothing was wrong.

File: sycl/include/sycl/ext/oneapi/bindless_images.hpp
Comment (maarquitos14) [reply sycl/include/sycl/ext/oneapi/bindless_images.hpp]: I'm good with it, as long as implementation matches spec.

File: sycl/include/sycl/ext/oneapi/bindless_images.hpp
Comment (sommerlukas) [reply sycl/include/sycl/ext/oneapi/bindless_images.hpp:1207]: @bader What ABI guarantees do we give for experimental extensions? 

https://intel.github.io/llvm-docs/developer/ABIPolicyGuide.html also talks about an ABI-breaking window coming up in April. Have the dates for this window already been finalized?

@Seanst98 If it was decided that we want to give ABI guarantees for experimental features, you could look into https://github.com/intel/llvm/pull/11629 for how to guard such changes between ABI breaking windows.

File: sycl/include/sycl/ext/oneapi/bindless_images.hpp
Comment (maarquitos14) [reply sycl/include/sycl/ext/oneapi/bindless_images.hpp:1192]: Yes, as long as it's being tracked, it works for me.

File: sycl/include/sycl/ext/oneapi/bindless_images.hpp
Comment (maarquitos14) [reply sycl/include/sycl/ext/oneapi/bindless_images.hpp:1207]: From the [spec](https://github.com/intel/llvm/blob/52d984fd910ebe6059397667fdd71d386ae82e51/sycl/doc/extensions/experimental/sycl_ext_oneapi_bindless_images.asciidoc#status)

> This is an experimental extension specification, intended to provide early access to features and gather community feedback. Interfaces defined in this specification are implemented in DPC++, but they are not finalized and may change incompatibly in future versions of DPC++ without prior notice. Shipping software products should not rely on APIs defined in this specification.

So I guess you can freely change the ABI.

Comment (aelovikov-intel) [reply sycl/include/sycl/ext/oneapi/bindless_images.hpp:1207]: > https://intel.github.io/llvm-docs/developer/ABIPolicyGuide.html also talks about an ABI-breaking window coming up in April. Have the dates for this window already been finalized?

Started today, AFAIK.

Comment (bader) [reply sycl/include/sycl/ext/oneapi/bindless_images.hpp:1207]: > From the [spec](https://github.com/intel/llvm/blob/52d984fd910ebe6059397667fdd71d386ae82e51/sycl/doc/extensions/experimental/sycl_ext_oneapi_bindless_images.asciidoc#status)
> 
> > This is an experimental extension specification, intended to provide early access to features and gather community feedback. Interfaces defined in this specification are implemented in DPC++, but they are not finalized and may change incompatibly in future versions of DPC++ without prior notice. Shipping software products should not rely on APIs defined in this specification.
> 
> So I guess you can freely change the ABI.

That's right. 👍 

ABI policy guidelines do not apply to experimental features. Here is the quote from the [ABI policy guide](https://intel.github.io/llvm-docs/developer/ABIPolicyGuide.html):

> Note: Features clearly marked as experimental are considered as an exception to this guideline.

About this comment.

> > https://intel.github.io/llvm-docs/developer/ABIPolicyGuide.html also talks about an ABI-breaking window coming up in April. Have the dates for this window already been finalized?
> 
> Started today, AFAIK.

@stdale-intel, can you confirm that by updating the guide, please? As of now it's not clear if ABI breaking changes are allowed.

> ABI breaking changes window April 18 - July 11 2023 [CLOSED]
> Next ABI breaking changes window is tenatively April 2024.

**Review by maarquitos14 — APPROVED** [review]

File: sycl/source/detail/device_impl.cpp
Comment (steffenlarsen) [reply sycl/source/detail/device_impl.cpp:588]: I believe there is a good reason for `call` rather than `call_nocheck` in this case. Historically, it was useful as a lot of the device info enumerators were 1:1 with OpenCL and were passed directly to the OpenCL implementation. If the implementations did not recognize the enum value, the backend would return an error code. Though that is unlikely to be how these enums are implemented in UR now, having error codes indicate that the backend doesn't support the feature is a conservative solution and could also let newer DPC++ implementations use older UR plugins (I don't think we consider that valid in general though.)

Comment (Seanst98) [reply sycl/source/detail/device_impl.cpp]: I've switched to using `call` rather than `call_nocheck`.

Comment (steffenlarsen) [reply sycl/source/detail/device_impl.cpp]: Sorry, I switched the two around, I meant that `call_nocheck` had a benefit over `call` here. I would use coffee as an excuse, but 1PM feels a little too late for that. Once again, I apologize for the confusion.

Comment (przemektmalon) [reply sycl/source/detail/device_impl.cpp:588]: Reviving this thread as I've found that using `call` breaks all E2E tests on platforms that don't support all of our aspects.

My [sampled fetch PR](https://github.com/intel/llvm/pull/12447) was failing CI until I switched back to using `call_nocheck`.

For the moment I'd prefer to revert this and use `call_nocheck` like all the other queries in the file. I think changing all the queries to use `call` and work properly should be done in a separate PR, as the mechanism for catching exceptions from the `call` needs to be looked at as well.

What are your thoughts @steffenlarsen ?

Comment (steffenlarsen) [reply sycl/source/detail/device_impl.cpp:588]: @przemektmalon - I noticed your comment on your other PR and noticed I had switched the two around when I made my previous comment. I'm sorry for the confusion, but I am with you on using `call_nocheck`.

**Review by aarongreig — APPROVED** [review]

**Comment by przemektmalon** [comment]
Ping @intel/llvm-gatekeepers can this be merged?

**Pull request merged**
